packages/core/src/browser/shell/tab-bar-toolbar/tab-bar-toolbar.tsx

```diff
@@ -221,17 +246,23 @@ export class TabBarToolbar extends ReactWidget {
         return this.commands.isToggled(command, this.current);
     }
 
+    protected evaluateWhenClause(whenClause: string | undefined): boolean {
+        return whenClause ? this.contextKeyService.match(whenClause) : true;
+    }
+
     protected executeCommand = (e: React.MouseEvent<HTMLElement>) => {
         e.preventDefault();
         e.stopPropagation();
 
         const item: AnyToolbarItem | undefined = this.inline.get(e.currentTarget.id);
-        if (item?.command && item.menuPath) {
-            this.menuCommandExecutor.executeCommand(item.menuPath, item.command, this.current);
-        } else if (item?.command) {
-            this.commands.executeCommand(item.command, this.current);
-        } else if (item?.menuPath) {
-            this.renderMoreContextMenu(this.toAnchor(e), item.menuPath);
+        if (this.evaluateWhenClause(item?.when)) {
+            if (item?.command && item.menuPath) {
+                this.menuCommandExecutor.executeCommand(item.menuPath, item.command, this.current);
+            } else if (item?.command) {
+                this.commands.executeCommand(item.command, this.current);
+            } else if (item?.menuPath) {
+                this.renderMoreContextMenu(this.toAnchor(e), item.menuPath);
+            }
```

Comment:
We can avoid all these changes, and the nested if statement by performing the opposite logic which is also clearer:

```ts
if (!this.evaluateWhenClause(item?.when)) {
    return;
}

...
```